test_conformance/math_brute_force/i_unary_half.cpp

```diff
@@ -0,0 +1,215 @@
+                              ParameterType::Half, vector_size_index);
+    };
+    return BuildKernels(info, job_id, generator);
+}
+
+} // anonymous namespace
+
+int TestFunc_Int_Half(const Func *f, MTdata d, bool relaxedMode)
+{
+    int error;
+    Programs programs;
+    KernelMatrix kernels;
+    const unsigned thread_id = 0; // Test is currently not multithreaded.
+    int ftz = f->ftz || 0 == (gHalfCapabilities & CL_FP_DENORM) || gForceFTZ;
+    size_t bufferSize = BUFFER_SIZE;
+    uint64_t step = getTestStep(sizeof(cl_half), BUFFER_SIZE);
+    uint64_t bufferElements = bufferSize / sizeof(cl_int);
+    std::vector<float> s(0);
+
+    int scale = (int)((1ULL << 16) / (16 * bufferElements) + 1);
+
+    logFunctionInfo(f->name, sizeof(cl_half), relaxedMode);
+    // This test is not using ThreadPool so we need to disable FTZ here
+    // for reference computations
+    FPU_mode_type oldMode;
+    DisableFTZ(&oldMode);
+    std::shared_ptr<int> at_scope_exit(
+        nullptr, [&oldMode](int *) { RestoreFPState(&oldMode); });
+
+    // Init the kernels
+    {
+        BuildKernelInfo build_info = { 1, kernels, programs, f->nameInCode };
+        if ((error = ThreadPool_Do(BuildKernel_HalfFn,
+                                   gMaxVectorSizeIndex - gMinVectorSizeIndex,
+                                   &build_info)))
+            return error;
+    }
+    s.resize(bufferElements);
+
+    for (uint64_t i = 0; i < (1ULL << 16); i += step)
+    {
+        // Init input array
+        cl_ushort *p = (cl_ushort *)gIn;
+        if (gWimpyMode)
+        {
+            for (size_t j = 0; j < bufferElements; j++)
+                p[j] = (cl_ushort)i + j * scale;
+        }
+        else
+        {
+            for (size_t j = 0; j < bufferElements; j++) p[j] = (cl_ushort)i + j;
+        }
+        if ((error = clEnqueueWriteBuffer(gQueue, gInBuffer, CL_FALSE, 0,
+                                          bufferSize, gIn, 0, NULL, NULL)))
+        {
+            vlog_error("\n*** Error %d in clEnqueueWriteBuffer ***\n", error);
+            return error;
+        }
+
+        // write garbage into output arrays
+        for (auto j = gMinVectorSizeIndex; j < gMaxVectorSizeIndex; j++)
+        {
+            uint32_t pattern = 0xffffdead;
+            memset_pattern4(gOut[j], &pattern, bufferSize);
+            if ((error =
+                     clEnqueueWriteBuffer(gQueue, gOutBuffer[j], CL_FALSE, 0,
+                                          bufferSize, gOut[j], 0, NULL, NULL)))
+            {
+                vlog_error("\n*** Error %d in clEnqueueWriteBuffer2(%d) ***\n",
+                           error, j);
+                return error;
+            }
+        }
+
+        // Run the kernels
+        for (auto j = gMinVectorSizeIndex; j < gMaxVectorSizeIndex; j++)
+        {
+            size_t vectorSize = sizeValues[j] * sizeof(cl_int);
+            size_t localCount = (bufferSize + vectorSize - 1) / vectorSize;
```

Comment:
Actually I think this calculation may be correct. In order to prevent access violation due to overflowing of the index of output buffer (which is the same size as input half buffer) we need to use size of bigger type to calculate work size of the kernel. Am I missing something here ?

Reply:
After some digging I believe the calculation may very well be correct, but the inputs to the calculation are probably too large.

For example, for the scalar case in non-wimpy mode: `bufferSize = 1024 * 1024 * 2`, which is `524288` `int`s, so we're enqueuing the kernel with a global work size of `524288`.  That is 8 times more than the 65536 possible fp16 input values.  So every possible fp16 input value is repeated 8 times, which makes the test unnecessarily slow (particularly noticeable on embedded platforms).

I guess this problem stems from the original (float/double) structure of the code, which doesn't consider the case when all possible input values can easily fit in a single buffer.

Reply:
@svenvh I just applied corrections for `bufferSize` in `i_unary_half.cpp`. Two remarks here:

-buffer size computation schema is common for most of `cl_khr_fp16` `bruteforce` tests, please let me know if this correction works for you so I could propagate it across remaining tests
-with this correction previous approach for `gWimpyMode` make no longer sense so I removed section to initialize buffer for wimpy mode, if you have other suggestions I will appreciate it

Reply:
> I just applied corrections for bufferSize in i_unary_half.cpp.

You only touched `bufferElements`, which doesn't affect the global work size that gets passed to `clEnqueueNDRangeKernel` as far as I can see.  So with your latest change the test still enqueues 8 times the amount of work needed, but only validates 1/8 of that work (i.e. it now validates the "correct" amount).

Reply:
Good catch, thanks. Uninitialized buffer passed the test which is odd. Size related variables were corrected.

Reply:
Thanks; with your latest correction I see the global work size is as expected. And the separation of buffer size in/out variables looks good to me.

Reply:
I found similar computation schema in `unary_two_results_half.cpp`, `unary_two_results_i_half.cpp` and `unary_u_half.cpp`. Relevant corrections applied.